### PR TITLE
Rework MaxNLocator, eliminating infinite loop; closes #6849

### LINKED ISSUE #6849: BUG: endless loop with MaxNLocator integer kwarg and short axis
In v2.x:
```python
fig, ax = plt.subplots()
ax.plot([0.5, 1.5], [1, 2])
ax.locator_params(axis='x', integer=True)
```
This leads to an infinite loop in `MaxNLocator._raw_ticks`.

### PR DESCRIPTION
**Pull request by efiring**
The core algorithm has been reimplemented in a more compact and understandable form than my earlier try at adding the `min_n_ticks` constraint.

I replaced the images for one mplot3d test because I think the present behavior is more consistent than the earlier behavior that led to those images.  In the replaced versions, all of the axes have ticks just slightly inside the limits.  Based on the gallery it appears this is the intended behavior but @weathergod can confirm or deny this.

### CODE REVIEW

**Comment by efiring** [comment]
Here also, the appveyor py 3.5 build failed early on; it's unrelated to this PR.

File: lib/matplotlib/ticker.py [hunk 1]
```diff
@@ -1662,6 +1662,13 @@ def set_params(self, **kwargs):
                     steps = list(steps)
                     steps.append(10)
                 self._steps = steps
+            # Make an extended staircase within which the needed
+            # step will be found.  This is probably much larger
+            # than necessary.
+            flights = (0.1 * np.array(self._steps[:-1]),
```
Comment (tacaswell) [inline lib/matplotlib/ticker.py]: :+1: to the naming

**Comment by anntzer** [comment]
This leads to `integer=True` not being respected in the following case:
```
ax = gca(); ax.locator_params(integer=True); ax.set_xlim(-.1, 1.1)
```
(or the example in #6849) even though the ticker could very well just use the two ticks at the two integer values.

**Comment by efiring** [comment]
@anntzer you are correct with respect to your example; it is now fixed and works as I intended.  The example in #6849, however, leaves room for only one integer tick, so it will relax the integer constraint (given the default min_n_ticks=2).

**Comment by tacaswell** [comment]
backported to v2.x as 9982e40

**Pull request merged**